### ISSUE TITLE
Error on using RB access callback

### ISSUE THREAD

**Issue by georgehristov** [Telaxus/EPESI#131]
When using access callback following statement is set `$crits_raw[$action2][] = $callback_ret;` in `Utils_RecordBrowserCommon::get_access` method.
For blocked fields further down the method it is expected that `$crits_raw[$action]` has rule ids as keys:
`foreach ($crits_raw[$action] as $rule_id => $c)` .
This causes error to be triggered at `$blocked_fields = $fields[$rule_id];` as in case of callback `$rule_id` is the numeric key.

**Comment by bukowskiadam** [comment]
I see - working on that right now

**Comment by georgehristov** [comment]
I also get incorrect behaviour on the record access when `$record` parameter is not `null`. The `$crits_raw` for the view action contains 2 entries: first one is empty crits and the second one is the correct crits. Somehow this causes the system to allow access to all records (not based on the crits only.

This is an example of the `$crits_raw`. I presume `Utils_RecordBrowserCommon::check_record_against_crits` returns `true` when called with empty crits but did not go so far with the checks yet: 

```
array (size=7)
  'view' => 
    array (size=2)
      13 => 
        object(Utils_RecordBrowser_Crits)[69]
          protected 'negation' => boolean false
          protected 'join_operator' => null
          protected 'component_crits' => 
            array (size=0)
              empty
          protected 'active' => boolean true
      14 => 
        object(Utils_RecordBrowser_Crits)[68]
          protected 'negation' => boolean false
          protected 'join_operator' => null
          protected 'component_crits' => 
            array (size=1)
              0 => 
                object(Utils_RecordBrowser_CritsSingle)[75]
                  protected 'field' => string 'id' (length=2)
                  protected 'value' => 
                    array (size=3)
                      1 => string '1' (length=1)
                      2 => string '2' (length=1)
                      31 => string '31' (length=2)
                  protected 'operator' => string '=' (length=1)
                  protected 'negation' => boolean false
                  protected 'raw_sql_value' => boolean false
                  protected 'active' => boolean true
          protected 'active' => boolean true
  'edit' => 
    array (size=0)
      empty
  'delete' => 
    array (size=0)
      empty
  'add' => 
    array (size=0)
      empty
  'print' => 
    array (size=0)
      empty
  'export' => 
    array (size=0)
      empty
  'selection' => 
    array (size=0)
      empty
```

**Comment by bukowskiadam** [comment]
Try *testing* branch right now - I've made two commits addressing this and related issue.
I've discovered that returning true in access callback didn't work as expected. It returned standard crits instead of allowing view all records.

**Comment by georgehristov** [comment]
Now I do not get the error any more but the functionality is still not there (as per my second post),
If `$raw_crits` works like white list this means they are "merged" with "OR" while the callback `$crits` are merged with 'AND'.
Therefore there is unexpected behavior.
Generally the callback functionality acts a bit reverse on the EPESI permissions functionality - it does not allow but it rather limits the access.
I have access to a recordset set by callback. In regular permissions the record access is granted and it is limited to some records in the callback.
This works very well with standard functionality - displaying records in RB, etc.
But when I call `Utils_RecordBrowserCommon::get_access` method from my code supplying `$record` parameter then the method grants access to records which are otherwise (and supposedly) not accessible.

**Comment by bukowskiadam** [comment]
Maybe you're right that something is wrong right now...
Access callback - may return crits. Crits from all access callbacks are merged with AND. Then such final crits are treated as one crit and merged with OR.

But I don't get why EPESI code handles your case properly and your code don't. Could you please share your code (parts of it) that doesn't work as expected? Especially access callbacks

**Comment by georgehristov** [comment]
There is not really much to share.
To simulate: 
- grant access to a recordset (e.g. 'rs_test') using the standard RB permissions
- add access callback for rs_test that returns crits limiting the access to some records (e.g. id=>1)
- try browsing the records - in my case all is fine: i see only rs_test record with id=1
- but when I use following code (check access for record id=2) I get an array of the field=>access pairs and all are `true`:
```
$access = Utils_RecordBrowserCommon::get_access('rs_test', 'view', Utils_RecordBrowserCommon::get_record('rs_test', 2));
```
var_dump of `$access`:
```
array (size=3)
  'field_1' => boolean true
  'field_2' => boolean true
  'field_3' => boolean true
```

**Comment by bukowskiadam** [comment]
I see. That's the reason of a bad merge crits [here](https://github.com/Telaxus/EPESI/blob/testing/modules/Utils/RecordBrowser/RecordBrowserCommon_0.php#L1852)
If I'll change this to use `OR`, then you won't be able to limit access. Custom access callback will only add a separate rule that's not defined in the GUI. But then it'll be fully working and it'll properly check the access.

I think that both of the options *expanding* and *limiting* rule **may** be useful. I've marked may, because we don't use this right now. I think that you're the one using it :wink:
To implement both we should refactor this code and mark which crits are limiting etc.
What do you think about this?

**Comment by georgehristov** [comment]
I am not sure why you refer to the merge crits [here](https://github.com/Telaxus/EPESI/blob/testing/modules/Utils/RecordBrowser/RecordBrowserCommon_0.php#L1852) as bad as it does the job properly.
I believe the fix now would be to also implement the 'AND' functionality for `$crits_raw` [here](https://github.com/Telaxus/EPESI/blob/testing/modules/Utils/RecordBrowser/RecordBrowserCommon_0.php#L1874)
Is there a particular reason that `$crits_raw` work as a white list?
Of course more options for access is better as long as it is intuitive. Present set up works for me perfectly except for this case which I came across now. And my access setup is quite complicated I believe. So there is no immediate necessity to extend functionality, it already does a lot.

**Comment by georgehristov** [comment]
Now I believe I fixed it.
So `$crits` has following functionality:
(granted_by_access_rule_1 OR granted_by_access_rule_2 OR granted_by_access_rule_3 ...) AND granted_by_access_callback_crits
With this fix `$raw_crits` has same functionality

**Comment by georgehristov** [comment]
Yes, you are correct. I also believe it is not intuitive.
But I believe there might be a quick solution which I will test and let you know:
- remove all grants on the GUI
- use only callback to limit the record (it will act as a grant with crits)
This way all crits merges will be with 'OR'

**Comment by georgehristov** [comment]
In order for this to work as I mentioned above the callback should return an array of action->crits similar to the `$crits` variable used in get_access method.

**Comment by georgehristov** [comment]
I have adjusted my additional permissions module and the concept of white list including the callback crits works fine:
granted_by_access_rule_1 OR granted_by_access_rule_2 OR granted_by_access_rule_3 ... OR granted_by_access_callback_crits

The only question of concept is if the crits from different callbacks will also be merged with OR or as presently with AND [here](https://github.com/Telaxus/EPESI/blob/testing/modules/Utils/RecordBrowser/RecordBrowserCommon_0.php#L1799)
Let me know so I can prepare pull request.

**Comment by bukowskiadam** [comment]
In fact this function should return two sets of crits. One to grant access,
merged with or, one to restrict access merged with and. Each callback
should have its own type (grant/restrict) stored in database. Then we'll
have full set of possibilities.

**Comment by georgehristov** [comment]
Yes, you 're right. It does not quite work for me without the limiting functionality.
Who will work on the final solution - you or me?

**Comment by bukowskiadam** [comment]
I'd like to make a new release in a couple of days without modifications we're talking about. Feel free if you'd like to do this or have some spare time. I think, that we should open a new issue as a feature request.
